tests/Basics/AsyncTest.cpp

```diff
@@ -0,0 +1,332 @@
+            }
+            handle.resume();
+          }
+        }) {}
+
+  ~ConcurrentNoWait() {
+    if (_thread.joinable()) {
+      await_suspend(std::noop_coroutine());
+      _thread.join();
+    }
+  }
+
+  std::mutex _mutex;
+  std::condition_variable_any _cv;
+  std::deque<std::coroutine_handle<>> _coro;
+
+  std::jthread _thread;
+};
+
+struct InstanceCounterValue {
+  InstanceCounterValue() { instanceCounter += 1; }
+  InstanceCounterValue(InstanceCounterValue const& o) { instanceCounter += 1; }
+  InstanceCounterValue(InstanceCounterValue&& o) noexcept {
+    instanceCounter += 1;
+  }
+
+  ~InstanceCounterValue() {
+    if (instanceCounter == 0) {
+      abort();
+    }
+    instanceCounter -= 1;
+  }
+  static std::size_t instanceCounter;
+};
+
+std::size_t InstanceCounterValue::instanceCounter = 0;
+
+struct CopyOnlyValue : InstanceCounterValue {
+  CopyOnlyValue(int x) : x(x) {}
+  CopyOnlyValue(CopyOnlyValue const& x) : x(x) {}
+
+  auto operator<=>(int y) const noexcept { return x <=> y; }
+  operator int() const noexcept { return x; }
+  int x;
+};
+
+struct MoveOnlyValue : InstanceCounterValue {
+  MoveOnlyValue(int x) : x(x) {}
+  MoveOnlyValue(MoveOnlyValue&& x) : x(x) {}
+  MoveOnlyValue(MoveOnlyValue const& x) = delete;
+
+  auto operator<=>(int y) const noexcept { return x <=> y; }
+  operator int() const noexcept { return x; }
+  int x;
+};
+
+}  // namespace
+
+template<typename>
+struct AsyncTest;
+
+template<typename WaitType, typename ValueType>
+struct AsyncTest<std::pair<WaitType, ValueType>> : ::testing::Test {
+  void SetUp() override { InstanceCounterValue::instanceCounter = 0; }
+
+  void TearDown() override {
+    EXPECT_EQ(InstanceCounterValue::instanceCounter, 0);
+  }
+
+  WaitType wait;
+};
+
+using MyTypes = ::testing::Types<
+    std::pair<NoWait, CopyOnlyValue>, std::pair<WaitSlot, MoveOnlyValue>,
+    std::pair<ConcurrentNoWait, MoveOnlyValue>,
+    std::pair<NoWait, CopyOnlyValue>, std::pair<WaitSlot, MoveOnlyValue>,
+    std::pair<ConcurrentNoWait, CopyOnlyValue>>;
+TYPED_TEST_SUITE(AsyncTest, MyTypes);
+
+using namespace arangodb;
+
+TYPED_TEST(AsyncTest, async_return) {
+  using ValueType = TypeParam::second_type;
+
+  auto a = [&]() -> async<ValueType> {
+    co_await this->wait;
+    co_return 12;
+  }();
+
+  this->wait.resume();
+  EXPECT_TRUE(a.valid());
+  auto awaitable = std::move(a).operator co_await();
+  EXPECT_FALSE(a.valid());
+  this->wait.await();
+  EXPECT_TRUE(awaitable.await_ready());
+  EXPECT_EQ(awaitable.await_resume(), 12);
+}
+
+TYPED_TEST(AsyncTest, async_return_move) {
+  using ValueType = TypeParam::second_type;
+
+  auto a = [&]() -> async<ValueType> {
+    co_await this->wait;
+    co_return 12;
+  }();
+
+  EXPECT_TRUE(a.valid());
+
+  auto b = std::move(a);
+  EXPECT_TRUE(b.valid());
+  EXPECT_FALSE(a.valid());
+
+  a = std::move(b);
+  EXPECT_TRUE(a.valid());
+  EXPECT_FALSE(b.valid());
+
+  this->wait.resume();
+  this->wait.await();
+}
```

Comment:
Do you want to add something along the lines of
```
  EXPECT_EQ(std::move(a).operator co_await().await_resume(), 12);
```
here?

Reply:
No, this test tests what happens if you drop the future without awaiting it.